Improve `next_cache_entry` cache traversal performance

vdye opened the pull request:
## Changes
- Update `next_cache_entry` to accept a "cached" starting position for index search
  - Because `next_cache_entry` searches the `src_index`, which shouldn't be modified in-place over the course of `unpack_trees`, the first search position for the next `next_cache_entry` call can be `last_position + 1`

This might also help with `git blame` performance, if it's using `unpack_trees` (CC: @ldennington).

## Performance
This change ultimately only had an effect on sparse index checkouts, as demonstrated by the performance test results:

```
Test                                              ms/vfs-2.33.0     HEAD                  
------------------------------------------------------------------------------------------
2000.2: git reset -- does-not-exist (full-v3)     0.79(0.38+0.30)   0.91(0.43+0.34) +15.2%
2000.3: git reset -- does-not-exist (full-v4)     0.80(0.38+0.29)   0.85(0.40+0.35) +6.2% 
2000.4: git reset -- does-not-exist (sparse-v3)   0.76(0.43+0.69)   0.44(0.08+0.67) -42.1%
2000.5: git reset -- does-not-exist (sparse-v4)   0.71(0.40+0.65)   0.41(0.09+0.65) -42.3%
```

The reason appears to be because, in full indexes, `mark_ce_used` will advance `cache_bottom` such that, in `next_cache_entry`, already-checked entries are skipped. In sparse indexes, however, the presence of sparse directory entries will block the `cache_bottom` from advancing, so already-checked cache entries are _not_ skipped in `next_cache_entry`. Since the cache tree still needs `cache_bottom` to not advance (per 17a1bb570bcd165a3dc1c28c441bee45a941bb12), the transient `hint` position lets `next_cache_entry` shortcut already-checked entries even for a sparse index.

vdye committed:
unpack-trees: improve performance of `next_cache_entry`

To find the first non-unpacked cache entry, `next_cache_entry` iterates through
index, starting at `cache_bottom`, to find the first cache entry. The
performance of this in full indexes is helped by `cache_bottom` advancing with
each invocation of `mark_ce_used` (called by `unpack_index_entry`). However,
the presence of sparse directories can prevent the `cache_bottom` from
advancing in a sparse index case, effectively forcing `next_cache_entry` to
search from the beginning of the index each time it is called.

The need to preserve `cache_bottom` for the sparse index is documented in
17a1bb570bc, so to get the benefit it provides in "shortcutting"
already-searched entries a separate `hint` position is used. The hint position
is set inside of `next_cache_entry` to `last_searched_position + 1`, allowing
full _and_ sparse index iterations to skip already-searched entries.  The
performance is significantly improved for the sparse index case based on the
`p2000` results for a `git reset` with a non-matching pathspec (heavily using
`next_cache_entry`):

Test          ms/vfs-2.33.0     HEAD
------------------------------------------------------
(full-v3)     0.79(0.38+0.30)   0.91(0.43+0.34) +15.2%
(full-v4)     0.80(0.38+0.29)   0.85(0.40+0.35) +6.2%
(sparse-v3)   0.76(0.43+0.69)   0.44(0.08+0.67) -42.1%
(sparse-v4)   0.71(0.40+0.65)   0.41(0.09+0.65) -42.3%

Signed-off-by: Victoria Dye <vdye@github.com>

derrickstolee commented:
> ## Changes
> * Update `next_cache_entry` to accept a "cached" starting position for index search
>   
>   * Because `next_cache_entry` searches the `src_index`, which shouldn't be modified in-place over the course of `unpack_trees`, the first search position for the next `next_cache_entry` call can be `last_position + 1`

I think this is a very promising direction, especially in the short term.

> ```
> Test                                              ms/vfs-2.33.0     HEAD                  
> ------------------------------------------------------------------------------------------
> 2000.2: git reset -- does-not-exist (full-v3)     0.79(0.38+0.30)   0.91(0.43+0.34) +15.2%
> 2000.3: git reset -- does-not-exist (full-v4)     0.80(0.38+0.29)   0.85(0.40+0.35) +6.2% 
> 2000.4: git reset -- does-not-exist (sparse-v3)   0.76(0.43+0.69)   0.44(0.08+0.67) -42.1%
> 2000.5: git reset -- does-not-exist (sparse-v4)   0.71(0.40+0.65)   0.41(0.09+0.65) -42.3%
> ```

This is great!
 
> The reason appears to be because, in full indexes, `mark_ce_used` will advance `cache_bottom` such that, in `next_cache_entry`, already-checked entries are skipped. In sparse indexes, however, the presence of sparse directory entries will block the `cache_bottom` from advancing, so already-checked cache entries are _not_ skipped in `next_cache_entry`. Since the cache tree still needs `cache_bottom` to not advance (per [17a1bb5](https://github.com/microsoft/git/commit/17a1bb570bcd165a3dc1c28c441bee45a941bb12)), the transient `hint` position lets `next_cache_entry` shortcut already-checked entries even for a sparse index.

This is a good point to indicate that `cache_bottom` is not advancing the same way with sparse directory entries. We might want to dig into that more, since maybe there is something that causes other problems in the future. It's _probable_ that the things I did around `cache_bottom` could be done better.

But let's make a note to explore `cache_bottom` later (I'm thinking October timeframe, still with time to respond before the 2.34 release), but this is an excellent short-term step.

derrickstolee commented on t/perf/p2000-sparse-operations.sh:
```diff
@@ -112,6 +112,7 @@ test_perf_on_all git commit -a -m A
 test_perf_on_all git checkout -f -
 test_perf_on_all git reset
 test_perf_on_all git reset --hard
+test_perf_on_all git reset -- does-not-exist
```
Perhaps this should be a `fixup!` on the commit that added the surrounding tests to the script. I imagine that the other commit could be grouped with the other `git reset` changes when we upstream them.

derrickstolee commented on unpack-trees.c:
```diff
@@ -665,17 +665,26 @@ static void mark_ce_used_same_name(struct cache_entry *ce,
+static struct cache_entry *next_cache_entry(struct unpack_trees_options *o, int *hint)
 {
 	const struct index_state *index = o->src_index;
 	int pos = o->cache_bottom;
 
+	if (hint && *hint > pos)
+		pos = *hint;
```
It looks like `hint` is never `NULL`, so the `if (hint)` checks can be removed.

You are right to keep the `if (*hint > pos)` here, since we want to use `cache_bottom` when possible.

vdye replied:
Good point - I was going to make it optional, but there wasn't really a downside to including it in other usage of `next_cache_entry`. Removed the null checks!

derrickstolee commented on unpack-trees.c:
```diff
@@ -1386,12 +1395,13 @@ static int unpack_callback(int n, unsigned long mask, unsigned long dirmask, str
 	/* Are we supposed to look at the index too? */
 	if (o->merge) {
+		int hint = -1;
 		while (1) {
 			int cmp;
 			struct cache_entry *ce;
 
 			if (o->diff_index_cached)
-				ce = next_cache_entry(o);
+				ce = next_cache_entry(o, &hint);
```
Now that I look closely at your implementation, this old code had _quadratic_ complexity! That helps me understand why having ~70,000 index entries still took 5 seconds to iterate.

I suppose I was expecting `hint` to be initialized with some information other than `-1`, but this seems to be a good idea regardless of any issues with the sparse index.

derrickstolee reviewed (approved):
I was worried that the changes here could be short-term and want to be removed in the future, but it seems like adding this "hint memory" to `next_cache_entry()` is a good thing to do, anyway. It avoids quadratic growth in the algorithm!

Just a couple nits.

vdye committed:
fixup! p2000: add new commands

Intended to test handling of cache entries outside `traverse_trees`
(accomplished here by having the pathspec match no cache entries).

Signed-off-by: Victoria Dye <vdye@github.com>

ldennington commented on unpack-trees.c:
```diff
@@ -665,17 +665,24 @@ static void mark_ce_used_same_name(struct cache_entry *ce,
+	if (*hint > pos)
+		pos = *hint;
+
 	while (pos < index->cache_nr) {
```
Sorry for the retroactive review comments - if it would be better to discuss on a call lmk!

I see `index->cache_nr` frequently but am having some trouble understanding it/finding docs on what it means (I think you mentioned "cache number," but I'm still not sure why that's significant.

derrickstolee replied:
`cache_nr` is the number of entries in the `cache` array.

In the Git codebase, you'll see a lot of `X_nr` and `X_alloc` representing that there are `X_nr` elements in an array currently allocated to have `X_alloc` entries.

ldennington replied:
Thanks!

ldennington commented on unpack-trees.c:
```diff
@@ -665,17 +665,24 @@ static void mark_ce_used_same_name(struct cache_entry *ce,
+static struct cache_entry *next_cache_entry(struct unpack_trees_options *o, int *hint)
 {
 	const struct index_state *index = o->src_index;
 	int pos = o->cache_bottom;
```
I know `cache_bottom` is critical to this change, but a little unsure what it means?

derrickstolee replied:
As we scan the index, we are also drilling down the trees, so the cache-tree extension (if available) is giving us some idea of the range of index entries corresponding to a given tree. This can be computed dynamically, too. The `cache_bottom` value is storing the expected lower-end of this range. As we recurse into subtrees, the `cache_bottom` is stored in a stack variable somewhere and then the `o->cache_bottom` value is rewritten as the recursion ends.

In general: the `cache_bottom` thing is necessary to make these index scans work properly, but it is complicated because the index is a flat list of files. It's just one more thing that is added on top to try and make sense of that difference with the trees in the object database.

vdye replied:
*Disclaimer: I'm still learning a lot of this too, so I might get some of this wrong*

`cache_bottom` is a position within the index as a whole used in `unpack_trees` to keep track of what has already been processed and added to the result, whether by just "unpacking" the index entry (copying it from source index -> result) or by merging it with the given tree(s). 

In non-sparse index cases, `cache_bottom` pretty much lines up with `hint` (and serves the same purpose - skipping all the already-unpacked index entries). However, for reasons I still need to fully understand, the `cache_bottom` is not incremented the same way (it stops at, I _think_ the first sparse directory position in the index). The result is that, without `hint`, `next_cache_entry` will start from the (early-position) `cache_bottom` and loop over entries it has already handled.
